### ISSUE TITLE
1RDM cloning is broken

### ISSUE THREAD

**Issue by jtkrogel** [QMCPACK/qmcpack#995]
As revealed by #994, the 1RDM works in the current code only when 1 OMP thread is used per MPI.  This indicates that cloning (per thread deep copy) of the DensityMatrices1B class, or perhaps it's contained SPOSet, is broken.

Fixing this issue is critical for large scale production runs involving the 1RDM.

**Comment by prckent** [comment]
General question: Is there a predictor for when deep copy is broken? e.g. Search for a specific call. This is unlikely to be the only broken estimator.

**Comment by jtkrogel** [comment]
The only way to really tell is to compare the output from an MPI-only run and an MPI+OMP run.

**Comment by ye-luo** [comment]
The AoS code requires many deep copy to handle pointers especially in the LCAO case. We should in generally move away from makeClone and rely on good copy constructors. Even if we have to use makeClone at the moment, try to use copy constructor inside.

**Comment by ye-luo** [comment]
The AoS code requires many deep copy to handle pointers especially in the LCAO case. We should in generally move away from makeClone and rely on good copy constructors. Even if we have to use makeClone at the moment, try to use copy constructor inside.

**Comment by jtkrogel** [comment]
I'm not sure how this is really different.  There will always be added code and interface functions needed to handle pointed to data.

# Master copy
Psi->spo = new spo
H->spo = Psi->spo

# cloning
Psi0 = Psi->makeClone() # Psi0->spo==spo0, a new spo copy
Psi1 = Psi->makeClone() # Psi1->spo==spo1

H0 = H.makeClone() # should not make a new copy of spo
H0->spo = Psi0->spo
H1 = H.makeClone()
H1->spo = Psi1.spo

I don't see how we can handle relinking cases like these (we are trying to deep copy a DAG by copying individual elements) with only pure copy constructors.

**Comment by ye-luo** [comment]
If there is a pointer, an explicit copy constructor is needed otherwise just rely on compiler generated copy constructor instead of maintaining explicit makeClone everywhere. The compiler does the whole DAG.

Copy constructor of Psi:
  spo = new spo_type(*spo_in);
spo_type has compiler generated copy constructor.

@jtkrogel In your example
Psi = new Psi(spo) on each thread is actually not guaranteed to have the identical Psi data as the master if the Psi on the master was modified after the initial construction.
Using more compiler generated copyConstructor helps consistency.

**Comment by jtkrogel** [comment]
This conversation really needs a whiteboard, so I will defer it until later.

**Comment by jtkrogel** [comment]
Back to the 1RDM issue.  I learned over the weekend that 1RDM cloning in older QMCPACK versions worked.  Since the 1RDM code has essentially not changed since 2014, the observed breakage in the current code must be due to changes (likely related to cloning) outside of this class.

Next I'll be tracking down when the breakage was introduced via bisection search (slow since each recompile requires first backporting the small changes made in #967).

**Comment by jtkrogel** [comment]
I've just learned that this bug is simpler and more benign overall than I had suspected previously.

The outputted density matrix depends on the number of threads, but differs from the correct density matrix only by a single scaling factor that depends on the number of threads.  The ratios of pairs of matrix elements are always correct.  Generally, the scaling factor increases with the number of threads, but the relationship is not linear:
```
trace(density matrix)
 1 thread :   4.00 (correct)
 2 threads:  78.73
 4 threads: 116.15
16 threads: 144.17
```
What exactly causes this remains a mystery.  Sadly the point of breakage is somewhere in a 7 month dead zone (June 2016 to Jan 2017) where the code fails to build on my test machine.

**Comment by jtkrogel** [comment]
There are actually two issues and they may (or may not) be connected.

An additional issue I discovered is that when using threads, the old code (e.g. SVN revistion 6366) only accumulates a non-zero contribution to the density matrix for one thread, resulting in the expected value being divided by the number of threads (and also with a larger error bar once the norm is corrected, confirming that only one thread contributes):
```
16x1  trace  4.0001(2)   (correct)
 4x4  trace  1.0000(1)   (divided by 4)
 1x16 trace  0.24993(7)  (divided by 16)
```

This behavior changes with git commit 2391098 (sorry @ye-luo), which sadly is a commit used to fix a nasty spline bug (see #40 and #49).  This commit only changes around 100 lines in a single file,  einspline_helper.hpp.  Following this commit, instead of contributing zero, addition of threads beyond the first results in nonsensical values for the norm (as noted above this amounts to a single number multiplying all density matrix elements, the 1RDM is otherwise fine):
```
16x1  trace     4.00   (correct)
 4x4  trace   231.26   (crazy)
 1x16 trace   288.09   (crazy)
```

@ye-luo any immediate thoughts on how commit 2391098 could do this?

**Comment by ye-luo** [comment]
The old code was normalizing the spline orbital. My commit fixed the rotation issue and abandon the normalization since the orbital from QE was normalized and the normalization based on a grid is never accurate.

**Comment by jtkrogel** [comment]
Resolved by #1509.  Closing.
